tower/src/limit/concurrency/service.rs

```diff
@@ -2,34 +2,38 @@ use super::future::ResponseFuture;
 
 use tower_service::Service;
 
-use super::sync::semaphore::{self, Semaphore};
 use futures_core::ready;
+use std::fmt;
+use std::future::Future;
+use std::mem;
+use std::pin::Pin;
 use std::sync::Arc;
 use std::task::{Context, Poll};
+use tokio::sync::{OwnedSemaphorePermit, Semaphore};
 
 /// Enforces a limit on the concurrent number of requests the underlying
 /// service can handle.
 #[derive(Debug)]
 pub struct ConcurrencyLimit<T> {
     inner: T,
-    limit: Limit,
+    semaphore: Arc<Semaphore>,
+    state: State,
 }
 
-#[derive(Debug)]
-struct Limit {
-    semaphore: Arc<Semaphore>,
-    permit: semaphore::Permit,
+enum State {
+    Waiting(Pin<Box<dyn Future<Output = OwnedSemaphorePermit> + 'static>>),
```

Comment:
```suggestion
    Waiting(Pin<Box<dyn Future<Output = OwnedSemaphorePermit> + Send + 'static>>),
```

Reply:
Thanks, that (obvious) addition gets it to build. Now to make it work…